compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala

```diff
@@ -2493,23 +2535,62 @@ abstract class GenJSCode extends plugins.PluginComponent
 
       // Prepare quadruplets of (formalArg, irType, tempVar, actualArg)
       // Do not include trivial assignments (when actualArg == formalArg)
-      val formalArgs = enclosingLabelDefParams(sym)
-      val actualArgs = args map genExpr
       val quadruplets = {
-        for {
-          (formalArgSym, actualArg) <- formalArgs zip actualArgs
-          formalArg = encodeLocalSym(formalArgSym)
-          if (actualArg match {
-            case js.VarRef(`formalArg`) => false
-            case _                      => true
-          })
-        } yield {
-          mutatedLocalVars += formalArgSym
-          val tpe = toIRType(formalArgSym.tpe)
-          (js.VarRef(formalArg)(tpe), tpe,
-              freshLocalIdent("temp$" + formalArg.name),
-              actualArg)
+        val formalArgs = enclosingLabelDefParams(sym)
+        val quadruplets =
+          List.newBuilder[(js.VarRef, jstpe.Type, js.Ident, js.Tree)]
+
+        for ((formalArgSym, arg) <- formalArgs.zip(args)) {
+          val formalArg = encodeLocalSym(formalArgSym)
+          val actualArg = genExpr(arg)
+
+          /* #3267 The formal argument representing the special `this` of a
+           * tailrec method can have the wrong type in the scalac symbol table.
+           * We need to patch it up, along with the actual argument, to be the
+           * enclosing class type.
+           * See the longer comment in genMethodDef() for more details.
+           *
+           * Note that only testing the `name` against `nme.THIS` is safe,
+           * given that `genStatOrExpr()` for `ValDef` asserts that no local
+           * variable named `nme.THIS` can happen, other than the ones
+           * generated for tailrec methods.
+           */
+          val isTailJumpThisLocalVar = formalArgSym.name == nme.THIS
+
+          val tpe =
+            if (isTailJumpThisLocalVar) currentClassType
+            else toIRType(formalArgSym.tpe)
+
+          val fixedActualArg = {
+            if (isTailJumpThisLocalVar &&
+                actualArg.tpe != tpe && tpe != jstpe.AnyType) {
+              /* See the comments in genMethodDef() for further details about
+               * the similar fixup.
+               */
+              actualArg match {
+                case js.AsInstanceOf(underlying, _) if underlying.tpe == tpe =>
+                  underlying
+                case _ =>
+                  js.AsInstanceOf(actualArg, tpe.asInstanceOf[jstpe.ClassType])
+              }
```

Comment:
Have you considered factoring this in a helper? Maybe `stripOrInsertAsInstanceOf`?